llvm/lib/RemoteCachingService/GRPCRelayCAS.cpp

```diff
@@ -0,0 +1,638 @@
+  return getInMemoryCASData(Handle).getDataSize();
+}
+
+Error GRPCRelayCAS::forEachRef(ObjectHandle Handle,
+                               function_ref<Error(ObjectRef)> Callback) const {
+  auto &Node = getInMemoryCASData(Handle);
+  for (const InMemoryIndexValueT *Ref : Node.getRefs())
+    if (Error E = Callback(toReference(*Ref)))
+      return E;
+  return Error::success();
+}
+
+ObjectRef GRPCRelayCAS::readRef(ObjectHandle Node, size_t I) const {
+  return toReference(*getInMemoryCASData(Node).getRefs()[I]);
+}
+
+size_t GRPCRelayCAS::getNumRefs(ObjectHandle Handle) const {
+  return getInMemoryCASData(Handle).getNumRefs();
+}
+
+ArrayRef<char> GRPCRelayCAS::getData(ObjectHandle Handle,
+                                     bool RequiresNullTerminator) const {
+  return getInMemoryCASData(Handle).getData();
+}
+
+static std::string casIDToString(CASID ID) {
+  std::string HashStr;
+  raw_string_ostream SS(HashStr);
+  SS << ID;
+  return HashStr;
+}
+
+Expected<CASID> GRPCRelayCAS::getCachedResult(CASID InputID) {
+  GetValueRequest Request;
+  ClientContext Context;
+  GetValueResponse Response;
+  Request.set_key(casIDToString(InputID));
+  grpc::Status Status = KVStub->GetValue(&Context, Request, &Response);
+  if (!Status.ok())
+    return createGRPCStatusError(Status);
+
+  auto Outcome = Response.outcome();
+  if (Outcome == GetValueResponse_Outcome_KEY_NOT_FOUND)
+    return createUnknownKeyError(InputID);
+
+  if (Response.has_error())
+    return createGRPCResponseError(Response.error());
+
+  std::string Value = Response.value().entries().at("CASID");
+  return parseID(Value);
+}
+
+Error GRPCRelayCAS::putCachedResult(CASID InputID, CASID OutputID) {
+  PutValueRequest Request;
+  ClientContext Context;
+  PutValueResponse Response;
+  std::string InputStr = casIDToString(InputID);
+  std::string OutputStr = casIDToString(OutputID);
+  // Store the CASID in the CASID dictionary field.
+  Request.mutable_value()->mutable_entries()->insert({"CASID", OutputStr});
+  Request.set_key(InputStr);
+  grpc::Status Status = KVStub->PutValue(&Context, Request, &Response);
+  if (!Status.ok())
+    return createGRPCStatusError(Status);
+
+  if (Response.has_error())
+    return createGRPCResponseError(Response.error());
+
+  return Error::success();
+}
+
+Expected<ObjectHandle>
+GRPCRelayCAS::storeFromOpenFileImpl(sys::fs::file_t FD,
+                                    Optional<sys::fs::file_status> FS) {
+  std::error_code EC;
+  sys::fs::mapped_file_region Map(FD, sys::fs::mapped_file_region::readonly,
+                                  FS->getSize(),
+                                  /*offset=*/0, EC);
+  if (EC)
+    return errorCodeToError(EC);
+
+  ArrayRef<char> Data(Map.data(), Map.size());
+  CASPutRequest Request;
+  ClientContext Context;
+  CASPutResponse Response;
+#ifdef __APPLE__
+  // FIXME: Move to FileSystem.h?
+  // Get file path from FD.
+  char Path[1024];
```

Comment:
nit: `s/1024/PATH_MAX/`?

Reply:
Sure, PATH_MAX is apple platform macro and do not exist on Windows. Since this is under `__APPLE__` PATH_MAX is fine I guess.